File: JavaScript/intro_sort.js
function insertionSort(arr, left, right) {
    for (let i = left + 1; i <= right; i++) {
        let key = arr[i];
        let j = i - 1;
        while (j >= left && arr[j] > key) {
            arr[j + 1] = arr[j];
            j--;
        }
        arr[j + 1] = key;
    }
}

function partition(arr, low, high) {
    let pivot = arr[high];
    let i = low - 1;
    for (let j = low; j < high; j++) {
        if (arr[j] < pivot) {
            i++;
            [arr[i], arr[j]] = [arr[j], arr[i]];
        }
    }
    [arr[i + 1], arr[high]] = [arr[high], arr[i + 1]];
    return i + 1;
}

function quickSort(arr, low, high, depthLimit) {
    if (low < high) {
        if (depthLimit === 0) {
            heapSort(arr, low, high);
            return;
        }
        let pi = partition(arr, low, high);
        quickSort(arr, low, pi - 1, depthLimit - 1);
        quickSort(arr, pi + 1, high, depthLimit - 1);
    }
}

function heapify(arr, n, i) {
    let largest = i;
    let left = 2 * i + 1;
    let right = 2 * i + 2;

    if (left < n && arr[left] > arr[largest]) {
        largest = left;
    }

    if (right < n && arr[right] > arr[largest]) {
        largest = right;
    }

    if (largest !== i) {
        [arr[i], arr[largest]] = [arr[largest], arr[i]];
        heapify(arr, n, largest);
    }
}

function heapSort(arr, low, high) {
    let n = high - low + 1;
    for (let i = Math.floor(n / 2) - 1; i >= 0; i--) {
        heapify(arr, n, i);
    }
    for (let i = n - 1; i > 0; i--) {
        [arr[0], arr[i]] = [arr[i], arr[0]];
        heapify(arr, i, 0);
    }
}

function introSort(arr) {
    let n = arr.length;
    let depthLimit = 2 * Math.floor(Math.log(n) / Math.log(2));
    quickSort(arr, 0, n - 1, depthLimit);
    insertionSort(arr, 0, n - 1);
}

let arr = [10, 7, 8, 9, 1, 5, 0];
introSort(arr);
console.log("Sorted array:", arr);

File: Kotlin/introSort.kt
fun introSort(arr: IntArray) {
    val depthLimit = 2 * log2(arr.size.toDouble()).toInt()
    introSort(arr, 0, arr.size - 1, depthLimit)
}

private fun introSort(arr: IntArray, start: Int, end: Int, depthLimit: Int) {
    if (end - start <= 16) {
        insertionSort(arr, start, end)
    } else if (depthLimit == 0) {
        heapSort(arr, start, end)
    } else {
        val pivot = partition(arr, start, end)
        introSort(arr, start, pivot - 1, depthLimit - 1)
        introSort(arr, pivot + 1, end, depthLimit - 1)
    }
}

private fun insertionSort(arr: IntArray, start: Int, end: Int) {
    for (i in start + 1..end) {
        val key = arr[i]
        var j = i - 1
        while (j >= start && arr[j] > key) {
            arr[j + 1] = arr[j]
            j--
        }
        arr[j + 1] = key
    }
}

private fun heapSort(arr: IntArray, start: Int, end: Int) {
    val n = end - start + 1
    for (i in n / 2 - 1 downTo start) {
        heapify(arr, n, i, start)
    }
    for (i in end downTo start + 1) {
        arr[start] = arr[i].also { arr[i] = arr[start] }
        heapify(arr, i - start, start, start)
    }
}

private fun heapify(arr: IntArray, n: Int, i: Int, start: Int) {
    var largest = i
    val left = 2 * (i - start) + 1 + start
    val right = 2 * (i - start) + 2 + start

    if (left < start + n && arr[left] > arr[largest]) {
        largest = left
    }
    if (right < start + n && arr[right] > arr[largest]) {
        largest = right
    }
    if (largest != i) {
        arr[i] = arr[largest].also { arr[largest] = arr[i] }
        heapify(arr, n, largest, start)
    }
}

private fun partition(arr: IntArray, low: Int, high: Int): Int {
    val pivot = arr[high]
    var i = low - 1
    for (j in low until high) {
        if (arr[j] <= pivot) {
            i++
            arr[i] = arr[j].also { arr[j] = arr[i] }
        }
    }
    arr[i + 1] = arr[high].also { arr[high] = arr[i + 1] }
    return i + 1
}

private fun log2(x: Double): Double {
    return Math.log(x) / Math.log(2.0)
}

fun main() {
    val arr = intArrayOf(10, 7, 8, 9, 1, 5, 0, -1, 3)
    introSort(arr)
    println("Sorted array: ${arr.joinToString(", ")}")
}
